### PR TITLE
Make FunctionBodyPtr a thin pointer.

### PR DESCRIPTION
**Pull request by nlewycky**
Also document Send+Sync for FunctionBodyPtr.

### CODE REVIEW

**Review by Hywan — APPROVED** [review]
I appreciate the simplification of the code. Thanks!

**Comment by Hywan** [comment]
bors try

File: lib/engine-object-file/src/artifact.rs [hunk 1]
```diff
@@ -319,10 +319,12 @@ impl ObjectFileArtifact {
 
             byte_buffer[0..WORD_SIZE]
                 .clone_from_slice(&bytes[cur_offset..(cur_offset + WORD_SIZE)]);
+            let fp = usize::from_ne_bytes(byte_buffer);
```
Comment (syrusakbary) [inline lib/engine-object-file/src/artifact.rs]: Can you explain what is going on here?

Comment (nlewycky) [reply lib/engine-object-file/src/artifact.rs]: `bytes` is a slice of bytes, at some offset (`cur_offset`) into that stream of bytes, there is a pointer. We grab the pointer out of that stream of bytes and ultimately cast it to FunctionBodyPtr.

I don't know why we're doing this, but this is what the code was already doing. I could look into why if you want?

File: lib/engine/src/trap/frame_info.rs [hunk 2]
```diff
@@ -227,18 +227,15 @@ impl Drop for GlobalFrameInfoRegistration {
 /// dropped, will be used to unregister all name information from this map.
 pub fn register(
     module: Arc<ModuleInfo>,
-    finished_functions: &BoxedSlice<LocalFunctionIndex, FunctionBodyPtr>,
+    finished_functions: &BoxedSlice<LocalFunctionIndex, (FunctionBodyPtr, usize)>,
```
Comment (syrusakbary) [inline lib/engine/src/trap/frame_info.rs:230]: Can we have a struct that owns both the pointer and the length? (rather than an unnamed tuple)

Comment (nlewycky) [reply lib/engine/src/trap/frame_info.rs]: Done

File: lib/engine-jit/src/engine.rs [hunk 3]
```diff
@@ -193,7 +193,7 @@ impl JITEngineInner {
         custom_sections: &PrimaryMap<SectionIndex, CustomSection>,
     ) -> Result<
         (
-            PrimaryMap<LocalFunctionIndex, FunctionBodyPtr>,
+            PrimaryMap<LocalFunctionIndex, (FunctionBodyPtr, usize)>,
```
Comment (syrusakbary) [inline lib/engine-jit/src/engine.rs:196]: Update this to use `FunctionExtent` ?

File: lib/engine-jit/src/link.rs [hunk 4]
```diff
@@ -8,20 +8,17 @@ use wasmer_compiler::{
...
 use wasmer_types::LocalFunctionIndex;
 use wasmer_vm::ModuleInfo;
-use wasmer_vm::{FunctionBodyPtr, SectionBodyPtr, VMFunctionBody};
+use wasmer_vm::{FunctionBodyPtr, SectionBodyPtr};
 
 fn apply_relocation(
     body: usize,
     r: &Relocation,
-    allocated_functions: &PrimaryMap<LocalFunctionIndex, FunctionBodyPtr>,
+    allocated_functions: &PrimaryMap<LocalFunctionIndex, (FunctionBodyPtr, usize)>,
```
Comment (syrusakbary) [inline lib/engine-jit/src/link.rs:16]: Update this to use `FunctionExtent` ?

File: lib/engine-jit/src/link.rs [hunk 5]
```diff
@@ -69,7 +65,7 @@ fn apply_relocation(
 /// required relocations and jump tables.
 pub fn link_module(
     _module: &ModuleInfo,
-    allocated_functions: &PrimaryMap<LocalFunctionIndex, FunctionBodyPtr>,
+    allocated_functions: &PrimaryMap<LocalFunctionIndex, (FunctionBodyPtr, usize)>,
```
Comment (syrusakbary) [inline lib/engine-jit/src/link.rs:68]: Update me!

File: lib/engine-jit/src/engine.rs (hunk 3 above)
Comment (nlewycky) [reply lib/engine-jit/src/engine.rs]: Right, sorry. Done.

File: lib/engine/src/trap/frame_info.rs [hunk 6]
```diff
@@ -219,6 +219,16 @@ impl Drop for GlobalFrameInfoRegistration {
     }
 }
 
+/// Represents a continuous region of executable memory starting with a function
+/// entry point.
+pub struct FunctionExtent {
```
Comment (MarkMcCaskey) [inline lib/engine/src/trap/frame_info.rs:224]: `#[derive(Debug)]` please!

Comment (MarkMcCaskey) [reply lib/engine/src/trap/frame_info.rs:224]: Does this need to be `repr(C)`? 🤔

File: lib/engine/src/trap/frame_info.rs [hunk 7]
```diff
@@ -227,18 +237,22 @@ impl Drop for GlobalFrameInfoRegistration {
...
+    for (
+        i,
+        FunctionExtent {
+            ptr: start,
+            length: len,
+        },
+    ) in finished_functions.iter()
+    {
+        let start = **start as usize;
+        let end = start + len;
```
Comment (MarkMcCaskey) [inline lib/engine/src/trap/frame_info.rs:255]: Nice improvement!

I think the layout of `*const [T]` isn't guaranteed to be `ptr, len` so this is definitely an improvement in correctness

File: lib/vm/src/lib.rs [hunk 8]
```diff
@@ -58,24 +58,26 @@ pub use crate::vmoffsets::{TargetSharedSignatureIndex, VMOffsets};
...
 
 /// # Safety
-/// TODO:
+/// The VMFunctionBody that this points to is opaque, so there's no data to
+/// read or write through this pointer. This is essentially a usize.
 unsafe impl Send for FunctionBodyPtr {}
 /// # Safety
-/// TODO:
+/// The VMFunctionBody that this points to is opaque, so there's no data to
+/// read or write through this pointer. This is essentially a usize.
```
Comment (MarkMcCaskey) [inline lib/vm/src/lib.rs:80]: It's surprising that `Send` isn't automatically implemented for types that are both `Sync` and `Copy`

**Review by MarkMcCaskey — APPROVED** [review]
Looks good overall to me!

File: lib/engine-object-file/src/artifact.rs (hunk 1 above)
Comment (MarkMcCaskey) [reply lib/engine-object-file/src/artifact.rs]: This code is parsing in metadata stored in the object file, it's creating types like `usize` from the buffer of bytes

File: lib/engine/src/trap/frame_info.rs (hunk 6 above)
Comment (nlewycky) [reply lib/engine/src/trap/frame_info.rs:224]: I don't think it needs to be repr(C) right now, but I'd like to use FunctionBodyPtr in the VM types so I'll go ahead and add it now.

**Pull request merged**